inc/Menu/Item/AbstractItem.php

```diff
@@ -0,0 +1,197 @@
+    const CTX_DESKTOP = 1;
+    /** menu item is to be shown on mobile screens only */
+    const CTX_MOBILE = 2;
+    /** menu item is to be shown in all contexts */
+    const CTX_ALL = 3;
+
+    /** @var string name of the action, usually the lowercase class name */
+    protected $type = '';
+    /** @var string optional keyboard shortcut */
+    protected $accesskey = '';
+    /** @var string the page id this action links to */
+    protected $id = '';
+    /** @var string the method to be used when this action is used in a form */
+    protected $method = 'get';
+    /** @var array parameters for the action (should contain the do parameter) */
+    protected $params = array();
+    /** @var bool when true, a rel=nofollow should be used */
+    protected $nofollow = true;
+    /** @var string this item's label may contain a placeholder, which is replaced with this */
+    protected $replacement = '';
+    /** @var string the full path to the SVG icon of this menu item */
+    protected $svg = DOKU_INC . 'lib/images/menu/00-default_checkbox-blank-circle-outline.svg';
+    /** @var string can be set to overwrite the default lookup in $lang.btn_* */
+    protected $label = '';
+    /** @var int the context this titme is shown in */
+    protected $context = self::CTX_ALL;
+
+    /**
+     * AbstractItem constructor.
+     *
+     * Sets the dynamic properties
+     *
+     * Children should always call the parent constructor!
+     *
+     * @throws \RuntimeException when the action is disabled
+     */
+    public function __construct() {
+        global $ID;
+        $this->id = $ID;
+        $this->type = strtolower(substr(strrchr(get_class($this), '\\'), 1));
+        $this->params['do'] = $this->type;
+
+        if(!actionOK($this->type)) throw new \RuntimeException("action disabled: {$this->type}");
+    }
+
+    /**
+     * Return this item's label
+     *
+     * When the label property was set, it is simply returned. Otherwise, the action's type
+     * is used to look up the translation in the main language file and, if used, the replacement
+     * is applied.
+     *
+     * @return string
+     */
+    public function getLabel() {
+        if($this->label !== '') return $this->label;
+
+        /** @var array $lang */
+        global $lang;
+        $label = $lang['btn_' . $this->type];
+        if(strpos($label, '%s')) {
+            $label = sprintf($label, $this->replacement);
+        }
+        if($label === '') $label = '[' . $this->type . ']';
+        return $label;
+    }
+
+    /**
+     * Return the link this item links to
+     *
+     * Basically runs wl() on $id and $params. However if the ID is a hash it is used directly
+     * as the link
+     *
+     * @see wl()
+     * @return string
+     */
+    public function getLink() {
+        if($this->id[0] == '#') {
+            return $this->id;
+        } else {
+            return wl($this->id, $this->params);
```

Comment:
Not sure where the second url-encoding happens, but this results in links like `http://127.0.0.1/~michael/old/dokuwiki/doku.php?id=start&amp;do=edit` for me.